ethcore/sync/src/light_sync/sync_round.rs

```diff
@@ -148,16 +148,11 @@ impl Fetcher {
 	// collect complete requests and their subchain from the sparse header chain
 	// into the ready set in order.
 	fn collect_ready(&mut self) {
-		loop {
-			let start_hash = match self.sparse.front() {
-				Some(first) => first.hash(),
-				None => break,
-			};
-
-			match self.complete_requests.remove(&start_hash) {
-				None => break,
+		while let Some(first) = self.sparse.pop_front() {
```

Comment:
I don't know this code so I don't know if it makes sense to change this code to speculatively `pop` and then re-adding if `complete_requests.remove` fails. Is the success path the most common?

Reply:
Good point, Andre!

I don't know either this was more a hack to get around having a reference and mutating at the same time!
I somehow convinced myself by 

> This queue has `O(1)` amortized inserts and removals from both ends of the container

But now I realize that the ```Header struct``` is big (~600 bytes) so I will revert this thanks :raised_hands: